sycl/test-e2e/bindless_images/array/read_write_unsampled_array.cpp

```diff
@@ -0,0 +1,467 @@
+// RUN: %t.out
+
+#include "../bindless_helpers.hpp"
+#include <iostream>
+#include <random>
+#include <sycl/sycl.hpp>
+#include <type_traits>
+
+static sycl::device dev;
+
+// Uncomment to print additional test information
+// #define VERBOSE_PRINT
+
+// Helpers and utilities
+struct util {
+  // parallel_for 3D
+  template <int NDims, typename DType, int NChannels, typename KernelName,
+            typename = std::enable_if_t<NDims == 3>>
+  static void run_ndim_test(
+      sycl::queue q, sycl::range<3> globalSize, sycl::range<3> localSize,
+      sycl::ext::oneapi::experimental::unsampled_image_handle input_0,
+      sycl::ext::oneapi::experimental::unsampled_image_handle input_1,
+      sycl::ext::oneapi::experimental::unsampled_image_handle output) {
+    using VecType = sycl::vec<DType, NChannels>;
+    try {
+      q.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<KernelName>(
+            sycl::nd_range<NDims>{globalSize, localSize},
+            [=](sycl::nd_item<NDims> it) {
+              size_t dim0 = it.get_global_id(0);
+              size_t dim1 = it.get_global_id(1);
+              size_t dim2 = it.get_global_id(2);
+
+              if constexpr (NChannels >= 1) {
+                VecType px1 =
+                    sycl::ext::oneapi::experimental::read_image_array<VecType>(
+                        input_0, sycl::int2(dim0, dim1), int(dim2));
+                VecType px2 =
+                    sycl::ext::oneapi::experimental::read_image_array<VecType>(
+                        input_1, sycl::int2(dim0, dim1), int(dim2));
+
+                auto sum = VecType(
+                    bindless_helpers::add_kernel<DType, NChannels>(px1, px2));
+                sycl::ext::oneapi::experimental::write_image_array<VecType>(
+                    output, sycl::int2(dim0, dim1), int(dim2), VecType(sum));
+              } else {
+                DType px1 =
+                    sycl::ext::oneapi::experimental::read_image_array<DType>(
+                        input_0, sycl::int2(dim0, dim1), int(dim2));
+                DType px2 =
+                    sycl::ext::oneapi::experimental::read_image_array<DType>(
+                        input_1, sycl::int2(dim0, dim1), int(dim2));
+
+                auto sum = DType(
+                    bindless_helpers::add_kernel<DType, NChannels>(px1, px2));
+                sycl::ext::oneapi::experimental::write_image_array<DType>(
+                    output, sycl::int2(dim0, dim1), int(dim2), DType(sum));
+              }
+            });
+      });
+    } catch (sycl::exception e) {
+      std::cout << "\tKernel submission failed! " << e.what() << std::endl;
+      exit(-1);
+    } catch (...) {
+      std::cout << "\tKernel submission failed!" << std::endl;
+      exit(-1);
+    }
+  }
+
+  // parallel_for 2D
+  template <int NDims, typename DType, int NChannels, typename KernelName,
+            typename = std::enable_if_t<NDims == 2>>
+  static void run_ndim_test(
```

Comment:
I think `run_ndim_test` can be fused into a single function that handles both 2D and 3D cases. you can take a look at https://github.com/intel/llvm/blob/ae24111fc96146b7daa8aea6d09325ddee76c927/sycl/test-e2e/bindless_images/vulkan_interop/unsampled_images.cpp#L144